lib/cloud/asset.js

```diff
@@ -47,3 +53,150 @@ export function staticAssetHandler(req) {
     body: data
   });
 }
+
+class Signer {
+  sign() {
+    throw new Error('Not implemented, subclass should override this method');
+  }
+}
+
+/* eslint-disable camelcase */
+class FSSigner extends Signer {
+  constructor(_settings) {
+    super();
+    this.settings = _settings;
+  }
+
+  sign(name) {
+    const prefix = this.settings.assetStoreURLPrefix;
+    const duration = this.settings.assetStoreURLExpireDuration;
+    const expire = Math.floor(Date.now() / 1000) + duration;
+    const secret = this.settings.assetStoreSecret;
+    const hash = crypto.createHmac('sha256', secret)
+      .update(name)
+      .update(expire.toString())
+      .digest('base64')
+      .substring(0, 40);
+    const encoded = URLSafeBase64.encode(hash);
+    const fullURL =
+      `${prefix}/${name}?expiredAt=${expire.toString()}&signature=${encoded}`;
+    return Promise.resolve(fullURL);
+  }
+}
+
+class CloudSigner extends Signer {
+  constructor(_settings) {
+    super();
+    this.settings = _settings;
+    this.signerToken = null;
+  }
+
+  refreshSignerToken() {
+    const appName = this.settings.appName;
+    const duration = this.settings.assetStoreURLExpireDuration;
+    const expire = Math.floor(Date.now() / 1000) + duration;
+    const token = settings.cloudAssetToken;
+    const host = this.settings.cloudAssetHost;
+    const url = `${host}/token/${appName}`;
+
+    return superagent.get(url)
+      .accept('application/json')
+      .set('Authorization', `Bearer ${token}`)
+      .query({expired_at: expire.toString()})
+      .then(response => {
+        const body = response.body;
+        console.log('fuck', body);
+        this.signerToken = {
+          value: body.value,
+          expiredAt: Date.parse(body.expired_at),
+          extra: body.extra
+        };
```

Comment:
Just put those data in `this`, not `this.signerToken`? I think an object should not need two/three level of internal data structure to make things complex.

And flat structure it save typing ;)